### PR TITLE
expression: add builtin function tidb_parse_tso

### PR DESCRIPTION
**Pull request by yu34po**
### What problem does this PR solve? 
https://github.com/pingcap/tidb/issues/8384

### What is changed and how it works?
add time builtin function tso 
 
### Check List 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

Related changes

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8385)

### CODE REVIEW

**Comment by jackysp** [comment]
Thanks for your contribution, @yu34po !
Please fix CI, seems the failure is about the timezone.

File: expression/builtin_time.go [hunk 1]
```diff
@@ -5629,3 +5630,50 @@ func getExpressionFsp(ctx sessionctx.Context, expression Expression) (int, error
...
+
+func (b *builtinTsoSig) Clone() builtinFunc {
+	newSig := &builtinTsoSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+// evalTime evals a builtinTsoSig.
+func (b *builtinTsoSig) evalTime(row chunk.Row) (types.Time, bool, error) {
+	arg, isNull, err := b.args[0].EvalInt(b.ctx, row)
```
Comment (dbjoa) [inline expression/builtin_time.go]: Why do we consider the case of ``isNull == true`` as an error ?

File: expression/builtin_time.go [hunk 2]
```diff
@@ -5629,3 +5630,50 @@ func getExpressionFsp(ctx sessionctx.Context, expression Expression) (int, error
...
+	if isNull || err != nil || arg <= 0 {
```
Comment (dbjoa) [inline expression/builtin_time.go]: Why do we consider the case of ``isNull == true`` as an error ?

Comment (yu34po) [reply expression/builtin_time.go]: function must accept a parameter

**Comment by dbjoa** [comment]
@yu34po The PR includes vender files, which should not be. Would you check whether or not you set the environment variable, `` GO111MODULE=on``?

**Comment by dbjoa** [comment]
@yu34po Would you also add the test for SQL statements having ``tidb_parse_tso()``? ``expression/integration_test.go`` can be a candidate place for the test.

File: expression/builtin.go [hunk 3]
```diff
@@ -413,6 +413,7 @@ var funcs = map[string]functionClass{
 	ast.Year:             &yearFunctionClass{baseFunctionClass{ast.Year, 1, 1}},
 	ast.YearWeek:         &yearWeekFunctionClass{baseFunctionClass{ast.YearWeek, 1, 2}},
 	ast.LastDay:          &lastDayFunctionClass{baseFunctionClass{ast.LastDay, 1, 1}},
+	ast.Tso:              &tsoFunctionClass{baseFunctionClass{ast.Tso, 1, 1}},
```
Comment (XuHuaiyu) [inline expression/builtin.go]: We'd better change ast.Tso to ast.TidbParseTso

File: expression/builtin_time.go [hunk 4]
```diff
@@ -5629,3 +5630,49 @@ func getExpressionFsp(ctx sessionctx.Context, expression Expression) (int, error
 	}
 	return mathutil.Min(expression.GetType().Decimal, types.MaxFsp), nil
 }
+
+type tsoFunctionClass struct {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: Please add a comment for the behavior of this function.

File: expression/builtin.go (hunk 3 above)
Comment (yu34po) [reply expression/builtin.go]: ast.Tso already merged in parser master

Comment (XuHuaiyu) [reply expression/builtin.go]: We can push another PR to change the name

**Comment by zz-jason** [comment]
@XuHuaiyu @dbjoa PTAL

File: expression/builtin_time.go (hunk 2 above)
Comment (dbjoa) [reply expression/builtin_time.go]: got it.

**Comment by dbjoa** [comment]
LGTM

File: expression/builtin_time.go [hunk 5]
```diff
@@ -5631,35 +5631,35 @@ func getExpressionFsp(ctx sessionctx.Context, expression Expression) (int, error
...
 }
 
-type builtinTsoSig struct {
+type builtintidbParseTsoSig struct {
 	baseBuiltinFunc
 }
 
-func (b *builtinTsoSig) Clone() builtinFunc {
-	newSig := &builtinTsoSig{}
+func (b *builtintidbParseTsoSig) Clone() builtinFunc {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/ builtintidbParseTsoSig/ builtinTidbParseTsoSig

File: go.mod [hunk 6]
```diff
@@ -49,3 +50,5 @@ require (
 	gopkg.in/natefinch/lumberjack.v2 v2.0.0
 	sourcegraph.com/sourcegraph/appdash v0.0.0-20180531100431-4c381bd170b4
 )
+
+replace github.com/pingcap/parser => github.com/yu34po/parser v0.0.0-20181123030610-1d6c3c585f2d
```
Comment (zz-jason) [inline go.mod]: please update `go.mod` to use the latest `pingcap/parser`.

**Comment by zz-jason** [comment]
@yu34po https://github.com/pingcap/parser/pull/50 has been merged, you can update `go.mod` to the latest `pingcap/parser` now.

File: go.mod [hunk 7]
```diff
@@ -41,6 +41,7 @@ require (
 	github.com/shirou/gopsutil v2.18.10+incompatible
 	github.com/sirupsen/logrus v1.2.0
 	github.com/spaolacci/murmur3 v0.0.0-20180118202830-f09979ecbc72
+	github.com/tiancaiamao/appdash v0.0.0-20181126055449-889f96f722a2
```
Comment (zz-jason) [inline go.mod]: why add this?

Comment (yu34po) [reply go.mod]: I don't know how this happened.
same as master expect parser version

Comment (zz-jason) [reply go.mod]: It disappears after merging the master branch..

**Pull request merged**